File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableDouble.java
package xyz.truenight.databinding.rxjava;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import androidx.databinding.ObservableDouble;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import xyz.truenight.utils.Optional;
import xyz.truenight.utils.Utils;

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

public class RxObservableDouble extends ObservableDouble {

    private Set<OnPropertyChangedCallback> count = new CopyOnWriteArraySet<>();

    private Disposable subscription;

    private Observable<Optional<Double>> observable;

    public RxObservableDouble(Observable<Optional<Double>> observable) {
        super();
        this.observable = observable;
    }

    public RxObservableDouble(double value, Observable<Optional<Double>> observable) {
        super(value);
        this.observable = observable;
    }

    public RxObservableDouble(BehaviorSubject<Optional<Double>> subject) {
        super(safe(subject.getValue()));
        this.observable = subject;
    }

    private static double safe(Optional<Double> value) {
        return Utils.safe(value == null ? null : value.orElse(null));
    }

    @Deprecated
    @Override
    public void set(double value) {
        super.set(value);
    }

    @Override
    public void addOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.addOnPropertyChangedCallback(callback);
        if (count.isEmpty()) {
            subscription = observable.subscribe(value -> super.set(safe(value)));
        }
        count.add(callback);
    }

    @Override
    public void removeOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.removeOnPropertyChangedCallback(callback);
        count.remove(callback);
        if (count.isEmpty()) {
            if (subscription != null && !subscription.isDisposed()) {
                subscription.dispose();
            }
        }
    }
}

File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableDouble.kt
package xyz.truenight.databinding.rxjava

import androidx.databinding.ObservableDouble
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import xyz.truenight.utils.optional.Optional
import xyz.truenight.utils.optional.safe
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

class RxObservableDouble @JvmOverloads internal constructor(val observable: Observable<Double>, default: Double, private val setter: ((Double) -> Unit)? = null) : ObservableDouble(default) {

    private val count = CopyOnWriteArraySet<androidx.databinding.Observable.OnPropertyChangedCallback>()

    private var subscription: Disposable? = null

    override fun set(value: Double) {
        super.set(value)
        setter?.let { it(value) }
    }

    override fun addOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.addOnPropertyChangedCallback(callback)
        if (count.isEmpty()) {
            subscription = observable.subscribe { super.set(it) }
        }
        count.add(callback)
    }

    override fun removeOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.removeOnPropertyChangedCallback(callback)
        count.remove(callback)
        if (count.isEmpty()) {
            subscription?.dispose()
        }
    }
}

private fun Optional<Double>?.safe() = this?.value.safe()

fun Observable<Double>.toBinding(default: Double = 0.0, setter: ((Double) -> Unit)? = null) =
        RxObservableDouble(this, default, setter)

fun Observable<Optional<Double>>.toBinding(default: Optional<Double> = Optional.empty(), setter: ((Double) -> Unit)? = null) =
        RxObservableDouble(this.map { it.safe() }, default.safe(), setter)

fun Flowable<Double>.toBinding(default: Double = 0.0, setter: ((Double) -> Unit)? = null) =
        RxObservableDouble(this.toObservable(), default, setter)

fun Flowable<Optional<Double>>.toBinding(default: Optional<Double> = Optional.empty(), setter: ((Double) -> Unit)? = null) =
        RxObservableDouble(this.map { it.safe() }.toObservable(), default.safe(), setter)